File: leet-code/937. Reorder Data in Log Files/937_reorder_data_in_log_files.py
def reorder_log_files(logs: list[str]) -> list[str]:
    letters = []
    digits = []

    for idx, log in enumerate(logs):
        if log.split()[1].isdigit():
            digits.append(log)
        else:
            letters.append(log)

    letters.sort(key=lambda x: (x.split()[1:], x.split()[0]))
    return letters + digits


print(reorder_log_files(["dig1 8 1 5 1","let1 art can","dig2 3 6","let2 own kit dig","let3 art zero"]))
print(reorder_log_files(["a1 9 2 3 1","g1 act car","zo4 4 7","ab1 off key dog","a8 act zoo"]))

File: leet-code/937. Reorder Data in Log Files/937_reorder_data_in_log_files.dart
class Solution {
  bool isdigit(String word) {
    RegExp regExp = RegExp(r'[0-9]');
    // https://api.flutter.dev/flutter/dart-core/RegExp-class.html
    return regExp.allMatches(word).toList().length == word.length;
  }

  List<String> reorderLogFiles(List<String> logs) {
    List<String> letters = [];
    List<String> digits = [];

    for (int i = 0; i < logs.length; i++) {
      String log = logs[i];
      String word = log.split(' ')[1];

      if (isdigit(word)) {
        digits.add(log);
      } else {
        letters.add(log);
      }
    }

    letters.sort((a, b) {
      String letterWithoutIdentifierA = a.substring(a.indexOf(' ') + 1);
      String letterWithoutIdentifierB = b.substring(b.indexOf(' ') + 1);

      // letter-logs 의 식별자를 제외한 content 가 같을 경우
      // 식별자로 정렬
      if (letterWithoutIdentifierA == letterWithoutIdentifierB) {
        return a.substring(0, a.indexOf(' ') + 1).compareTo(b.substring(0, b.indexOf(' ') + 1));
      }

      return letterWithoutIdentifierA.compareTo(letterWithoutIdentifierB);
    });

    return [...letters, ...digits];
  }
}

void main() {
  Solution solution = Solution();
  print(solution.reorderLogFiles(["dig1 8 1 5 1", "let1 art can", "dig2 3 6", "let2 own kit dig", "let3 art zero"]));
}

// 참고
// https://monsters-dev.tistory.com/39
